File: haskell/recv.hs
{-# LANGUAGE OverloadedStrings #-}

module Main where

import Control.Monad.Cont

import qualified Data.ByteString as B
import Network.Socket
import qualified Network.Socket.ByteString as B

dfltDest = tupleToHostAddress (127, 0, 0, 1)
dfltPort = 4321

main = withSocketsDo $ do
            -- Create the socket
            sock <- socket AF_INET Datagram 0

            -- Bind
            let addr = SockAddrInet dfltPort dfltDest
            bind sock addr

            -- Receive a message
            putStrLn "Awaiting 3 messages:"
            forM_ [1,2,3] $ \_ -> do
                resp <- B.recv sock 8192
                putStr "  Received \""
                B.putStr resp
                putStrLn "\""

            -- Close the socket
            close sock

File: python/recv.py
#!/usr/bin/env python3

import os
import sys
import socket

sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM, 0)
sock.bind( ('0.0.0.0', 4321) )

print("Awaiting 3 messages:")
for i in range(3):
    data = sock.recv(8192)
    print(f"  Received \"{data.decode('utf-8')}\"")
sock.close()
